CMakeLists.txt

```diff
@@ -263,215 +235,52 @@ if(MSVC AND NOT DEFINED CXX_FLAGS_MSVC)
+
 # MSVC's -Wall is not like gcc's, it really enables *all* warnings which include zillions for system headers and doesn't make sense.
 if(NOT MSVC)
-check_compiler_has_flag(CXX_FLAGS_PROJECT "-Wall" HAS_COMPILER_FLAG_WALL)
+	set(CXX_FLAGS_PROJECT "${CXX_FLAGS_PROJECT} -Wall")
 endif(NOT MSVC)
 
-### Set strict compiler flags.
-
-set(CXX_FLAGS_STRICT_COMPILATION)
-check_compiler_has_flag(
-	CXX_FLAGS_STRICT_COMPILATION
-	"-Werror"
-	HAS_COMPILER_FLAG_WERROR
-)
-
-# The current unit test code breaks strict aliasing with g++ 4.4.
-check_compiler_has_flag(
-	CXX_FLAGS_STRICT_COMPILATION
-	"-Wstrict-aliasing"
-	HAS_COMPILER_FLAG_WERROR_STRICT_ALIASING
-	"-Wno-strict-aliasing"
-)
-
-# This flag is/will be added in gcc-4.8 and fails with BOOST_STATIC_ASSERT
-check_compiler_has_flag(
-	CXX_FLAGS_STRICT_COMPILATION
-	"-Wunused-local-typedefs"
-	HAS_COMPILER_FLAG_WUNUSED_LOCAL_TYPEDEFS
-	"-Wno-unused-local-typedefs"
-)
-
-# This flag is/will be added in gcc-4.8 and fails with png in C++11 mode
-check_compiler_has_flag(
-	CXX_FLAGS_STRICT_COMPILATION
-	"-Wliteral-suffix"
-	HAS_COMPILER_FLAG_WLITERAL_SUFFIX
-	"-Wno-literal-suffix"
-)
-
-# This flag is too aggressive and keeps giving false positives.
-check_compiler_has_flag(
-	CXX_FLAGS_STRICT_COMPILATION
-	"-Wmaybe-uninitialized"
-	HAS_COMPILER_FLAG_WMAYBE_UNINITIALIZED
-	"-Wno-maybe-uninitialized"
-)
-
-# This removes a lot of warnings from Clang regarding unused -I arguments
-check_compiler_has_flag(
-	CXX_FLAGS_STRICT_COMPILATION
-	"-Qunused-arguments"
-	HAS_COMPILER_FLAG_QUNUSED_ARGUMENTS
-)
-
-# Silences Clang warnings about declaring a class a class first and
-# a struct later.
-check_compiler_has_flag(
-	CXX_FLAGS_STRICT_COMPILATION
-	"-Wmismatched-tags"
-	HAS_COMPILER_FLAG_WMISMATCHED_TAGS
-	"-Wno-mismatched-tags"
-)
-
-check_compiler_has_flag(
-	CXX_FLAGS_STRICT_COMPILATION
-	"-Wnull-conversion"
-	HAS_COMPILER_FLAG_WNULL_CONVERSION
-	"-Wno-null-conversion"
-)
-
-if(NOT CMAKE_COMPILER_IS_GNUCXX)
-	# Silences warnings about overloaded virtuals.
-	# (GCC doesn't complain Clang 3.2 does. Clang 3.4 no longer does.)
-	check_compiler_has_flag(
-		CXX_FLAGS_STRICT_COMPILATION
-		"-Woverloaded-virtual"
-		HAS_COMPILER_FLAG_WOVERLOADED_VIRTUAL
-		"-Wno-overloaded-virtual"
-	)
-endif(NOT CMAKE_COMPILER_IS_GNUCXX)
+set(COMPILER_FLAGS "${CXX_FLAGS_PROJECT}")
 
-check_compiler_has_flag(
-	CXX_FLAGS_STRICT_COMPILATION
-	"-Wold-style-cast"
-	HAS_COMPILER_FLAG_WOLD_STYLE_CAST
-)
+### Set strict compiler flags.
 
-check_compiler_has_flag(
-	CXX_FLAGS_STRICT_COMPILATION
-	"-Wdeprecated-register"
-	HAS_COMPILER_FLAG_WDEPRECATED_REGISTER
-	"-Wno-deprecated-register"
-)
+if(ENABLE_STRICT_COMPILATION)
+	set(CXX_FLAGS_STRICT_COMPILATION "-Werror -Wno-unused-local-typedefs -Wno-maybe-uninitialized -Wold-style-cast")
+	
+	if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+		set(CXX_FLAGS_STRICT_COMPILATION "${CXX_FLAGS_STRICT_COMPILATION} -Wmismatched-tags -Wno-literal-suffix -Wconditional-uninitialized")
```

Comment:
What's `-Wliteral-suffix`? I didn't find such a warning here: https://clang.llvm.org/docs/DiagnosticsReference.html

Reply:
It's defined for GCC [here](https://gcc.gnu.org/onlinedocs/gcc/C_002b_002b-Dialect-Options.html), but I added to to clang since it's used by travis+clang [here](https://github.com/wesnoth/wesnoth/blob/master/.travis.yml#L52).  Personally, it seems kind of odd for travis to be adding in its own warning/warning suppression options at all, but I didn't want to get into reviewing travis' options as well for this(except for getting rid of that `register` warning, since that seemed more important). I can remove it from both though, if you think that's correct.

Reply:
Please, do, review Travis' options

Reply:
Honestly, [neither of these checks](https://github.com/wesnoth/wesnoth/blob/master/.travis.yml#L47-L52) make a whole lot of sense to me - both check of OPT is -O0, then if it is, also add -O0 to EXTRA_FLAGS_RELEASE.

Reply:
It also looks like `-Wno-deprecated-declarations` would be added twice - once on line 49 and again for clang builds on line 52.

Reply:
Well, since Clang doesn't even support `-Wliteral-suffix` according to its documentation, it shouldn't be passed `-Wno-literal-suffix` in either `CMakeLists.txt` or `.travis.yml`.

Reply:
Well, in that case, I guess right now I'm confused what the OPT variable even does.  Aside from build 1, which uses `-O3 -O0` like I'd expect, all the other builds(ignoring the NLS build) just use straight `-O3`.

Otherwise, `-Wno-deprecated-declarations` could be added to the `strict` build for GCC and clang if that's something that should be suppressed, and `-Wno-literal-suffix` can be removed from the travis build options since it's already added by cmake/scons for clang with the PR, and at that point line 52 in .travis.yml can be removed.

Reply:
@jyrkive I'll just get rid of that flag entirely then.

Reply:
Another oddity, actually: cmake adds `-Wno-deprecated-declarations` [when Fribidi is enabled](https://github.com/wesnoth/wesnoth/blob/master/src/CMakeLists.txt#L142-L152), while scons [does it for lua](https://github.com/wesnoth/wesnoth/blob/master/src/SConscript#L113).